### PR TITLE
Feat/cargo workspace

### PR DESCRIPTION
**Pull request by spikespaz**
TODO: prune dependencies

### CODE REVIEW

**Comment by spikespaz** [comment]
I spent a lot of time making `cargo-udeps` work to only end up with two false-positive results. This is because while the crates are split, there is still much coupling. We can prune dependencies after shifting API surface.

**Comment by spikespaz** [comment]
Also will make a PR to set up [crane.dev](https://crane.dev)

**Comment by kuviman** [comment]
I don't understand why flake needed to be changed and why install nightly stuff

**Comment by kuviman** [comment]
Also I prefer root to be the main crate instead of it being nested, no real need to make a workspace, can just extract the module into a crate that's it

**Comment by spikespaz** [comment]
Well I hate all the flake templates I see as examples absolutely everywhere. They pull in tons of deps and are still cumbersome to work with. I wanted to standardize for myself, with feedback of course.

I am preparing to try https://crane.dev, and this project has a complicated dependency graph that could benefit from incremental builds. If you're planning on accepting pull requests without breakages from other viewers, best have CI help us know if it's ready. For that, [https://github.com/DeterminateSystems/magic-nix-cache](magic-nix-cache).

> I don't understand why flake needed to be changed and why install nightly stuff

Because `rustfmt` can't organize imports without nightly (will add config for that), and Rust-Analyzer has nightly-only features.

> Also I prefer root to be the main crate instead of it being nested, no real need to make a workspace, can just extract the module into a crate that's it

A workspace is required for shared development dependencies (TODO). It is also designed with parallel development in mind, by Cargo. When you think of a name for `twitch-bot` and it has a `publish = true` worthy API, we can just `filter-repo` the crates apart and preserve separate history.

**Comment by spikespaz** [comment]
Also if you meant `/kuvibot/ttv` and not `/{kuvibot,ttv}`, I want to depend on this for a twitch bot (workspace required for `= { git =`)

(the ttv name is taken and it confused me btw)

**Comment by kuviman** [comment]
This turns into a huge pr and I don't think I agree with all of it. I think it's better split it into smaller ones and discuss separately

**Comment by spikespaz** [comment]
I thought this was a small PR but ok what do you want to keep from this

**Comment by spikespaz** [comment]
I mean why not just edit this PR, you can. But please don't remove the nightly stuff from rust-overlay, I need RA nightly

**Comment by kuviman** [comment]
Ok not huge but there's multiple thing happening is what I mean

The most important part is moving ttv mod into twitch-bit crate and other things are irrelevant to it

When splitting the crate into multiple you don't need to move the root into a different folder but keep it in root which is my preference. This way you can still just cargo run from root or have examples in root if this was a library instead of searching through the repo

**Comment by spikespaz** [comment]
Ok I am awake and about to do your changes requested, but really want to make sure that your nested crate thing will still let me use `twitch_bot = { git = "https://github.com/kuviman/kuvibot" }` (it resolves the sub-path), because I don't think so. I am pretty sure nested crates are intended to be private, for proc macros and stuff, not for a sister library.
I don't think that a nested crate is even semantically correct here.

Actually there is something you can do to make it work. The outer `Cargo.toml` can have a `[package]` section as well as a `[workspace] members =`. The member names will still be exported and visible b Cargo from the root of the repository, but the directory structure can be what you want.

I really don't understand what the point is though, this was just supposed to be a small detail commit, not a big decision.

**Comment by spikespaz** [comment]
I'm aware that there are other ways to do it [but The Book](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html) tells us to use a workspace for this exact purpose.

**Comment by spikespaz** [comment]
Commands that work with the workspace:
```
cargo run
cargo run -p kuvibot
cargo build -p kuvibot
cargo build -p twitch-bot
```

**Comment by spikespaz** [comment]
> Ok not huge but there's multiple thing happening is what I mean

There's one thing happening in this PR and that is setting up a fairly normal Nix+Rust workspace. If you would like, we can separate the Nix and the Rust flake, but for reasons that I will elaborate if you ask, I literally *require* nightly with my current config.

This PR is just a prerequisite for other changes that I want to continue working on, so getting this (minimal) PR is foundational.

> The most important part is moving ttv mod into twitch-bit crate and other things are irrelevant to it

The `ttv` crate (renamed because of ambiguity, pick another one, `twitch-bot` didn't take much thought) is a `lib` and *definitely* belongs on the inside. That is the problem. A library crate should not contain a binary, but the other way around.

We can do this instead:

```toml
[package]
name = "mything"
version = "0.0.1"
authors = ["me <me@gmail.com>"]

[lib]
name = "mylib"
path = "src/lib.rs"

[[bin]]
name = "mybin"
path = "src/bin.rs"
```

This would make it one crate and maintain a sibling relationship between the `lib` and the `bin` (but under the same name which is wrong).

> When splitting the crate into multiple you don't need to move the root into a different folder but keep it in root which is my preference.

I've been doing more reading and your preference is incorrect. Nesting crates is usually only useful for re-exporting macros and proc-macros.

> This way you can still just cargo run from root or have examples in root if this was a library instead of searching through the repo

You can still do that, I'll move the example config to the root (but it can't be named `config.toml` at the root).

**Comment by spikespaz** [comment]
Another reason for using a workspace is to generate separate cargo doc for each package, but still have it navigable and built together.

**Comment by spikespaz** [comment]
Rebase split: https://github.com/kuviman/kuvibot/pull/2

**Comment by spikespaz** [comment]
Rebase split: #3

**Comment by kuviman** [comment]
Ok I pushed what I meant

This is still a workspace, but not a virtual workspace. The `[workspace]` section in the root is not even necessary, but I guess it helps with not creating a target folder in `twitch-bot` if you run cargo from there.

Otherwise, everything will work like you want to. You can use `twitch-bot = { git = "https://github.com/spikespaz/kuvibot", branch = "feat/cargo-workspace" }` to check.

File: src/lib.rs
Comment (kuviman) [inline src/lib.rs:1]: what is this doing here, its not even being used?

Comment (spikespaz) [reply src/lib.rs:1]: My bad will rm

**Comment by spikespaz** [comment]
not ready for re-review

**Comment by spikespaz** [comment]
Now it is ready

**Pull request merged**